listener.go

```diff
@@ -133,31 +133,14 @@ func (l *listener) Close() error {
 	l.closed = true
 	err := l.Listener.Close()
 
-	// Drain nextConn
-drainNextConn:
-	for {
-		select {
-		case conn := <-l.nextConn:
-			conn.Close()
-		default:
-			break drainNextConn
-		}
-	}
 	close(l.nextConn)
+	for conn := range l.nextConn {
+		conn.Close()
+	}
```

Comment:
`for range` loop is aware of closed channels, so the loop won't block. See the last section of https://dave.cheney.net/2014/03/19/channel-axioms

Reply:
this can also blow up actually I think (and so can the prior ones under the right conditions) -- it closes l.nextConn unilaterally, but there may still be potential writers on other goroutines, which would be a panic.  I think generally only-writers-close is the norm.

Reply:
I think these prior versions accomplish more of a best effort shutdown (but don't necessarily always work) -- They don't actually wait for the channel to close, they just wait until there is a point where there is nothing more to read immediately.